File: app/src/main/java/com/cyl/musiclake/ui/music/search/SearchPresenter.java
package com.cyl.musiclake.ui.music.search;

import com.cyl.musiclake.api.MusicApi;
import com.cyl.musiclake.api.MusicApiServiceImpl;
import com.cyl.musiclake.base.BasePresenter;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.bean.SearchHistoryBean;
import com.cyl.musiclake.net.ApiManager;
import com.cyl.musiclake.net.RequestCallBack;
import com.cyl.musiclake.utils.LogUtil;

import org.litepal.crud.DataSupport;
import org.litepal.crud.callback.FindMultiCallback;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by yonglong on 2018/1/6.
 */

public class SearchPresenter extends BasePresenter<SearchContract.View> implements SearchContract.Presenter {

    @Inject
    public SearchPresenter() {
    }

    @Override
    public void search(String key, SearchEngine.Filter type, int limit, int page) {
        ApiManager.request(MusicApiServiceImpl.INSTANCE
                        .searchMusic(key, type, limit, page)
                        .compose(mView.bindToLife()),
                new RequestCallBack<List<Music>>() {
                    @Override
                    public void success(List<Music> result) {
                        mView.showSearchResult(result);
                        mView.hideLoading();
                    }

                    @Override
                    public void error(String msg) {
                        mView.showEmptyView();
                        mView.hideLoading();
                    }
                });
    }

    @Override
    public void getSuggestions(String query) {
        DataSupport.select("title").where("title like ?", "%" + query + "%")
                .order("id desc").findAsync(SearchHistoryBean.class).listen(new FindMultiCallback() {
            @Override
            public <T> void onFinish(List<T> t) {
                mView.showSearchSuggestion((List<SearchHistoryBean>) t);
            }
        });
    }

    @Override
    public void saveQueryInfo(String query) {
        List<SearchHistoryBean> data = DataSupport.where("title = ?", query).find(SearchHistoryBean.class);
        if (data.size() == 0) {
            SearchHistoryBean historyBean = new SearchHistoryBean(System.currentTimeMillis(), query);
            if (historyBean.save()) {
                LogUtil.d("存储成功");
            } else {
                LogUtil.d("失败");
            }
        }
    }

    @Override
    public void getMusicInfo(int type, Music music) {
        if (music.getType() == Music.Type.QQ || music.getType() == Music.Type.NETEASE) {
            MusicApi.getMusicInfo(music)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Observer<Music>() {
                        @Override
                        public void onSubscribe(Disposable d) {

                        }

                        @Override
                        public void onNext(Music music) {
                            mView.showMusicInfo(type, music);
                        }

                        @Override
                        public void onError(Throwable e) {

                        }

                        @Override
                        public void onComplete() {

                        }
                    });
        } else {
            mView.showMusicInfo(type, music);
        }
    }


}

File: app/src/main/java/com/cyl/musiclake/ui/music/search/SearchPresenter.kt
package com.cyl.musiclake.ui.music.search

import com.cyl.musicapi.netease.SearchInfo
import com.cyl.musiclake.MusicApp
import com.cyl.musiclake.api.music.MusicApiServiceImpl
import com.cyl.musiclake.api.music.baidu.BaiduApiServiceImpl
import com.cyl.musiclake.api.music.netease.NeteaseApiServiceImpl
import com.cyl.musiclake.api.net.ApiManager
import com.cyl.musiclake.api.net.RequestCallBack
import com.cyl.musiclake.api.youtube.YoutubeDataApi
import com.cyl.musiclake.bean.HotSearchBean
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.SearchResult
import com.cyl.musiclake.data.db.DaoLitepal
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.ui.base.BasePresenter
import com.cyl.musiclake.utils.LogUtil
import com.google.api.services.youtube.model.SearchListResponse
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import javax.inject.Inject

/**
 * Created by yonglong on 2018/1/6.
 */

class SearchPresenter @Inject
constructor() : BasePresenter<SearchContract.View>(), SearchContract.Presenter {

    /**
     * 搜索Youtube返回的结果
     */
    fun searchByYouTube(query: String, pageToken: String, success: (SearchResult) -> Unit, fail: () -> Unit) {
        doAsync {
            val startTime = System.currentTimeMillis()
            val result: SearchListResponse? = YoutubeDataApi.search(query, pageToken)
            uiThread {
                result?.let {
                    val searchResult = SearchResult()
                    val songs = mutableListOf<Music>()
                    result.items.mapTo(songs) { t ->
                        val song = Music()
                        song.title = t.snippet.title
                        song.artist = t.snippet.channelTitle
                        song.artistId = t.snippet.channelId
                        song.mid = t.id.videoId
                        song.coverUri = t.snippet.thumbnails.high.url
                        song.type = Constants.YOUTUBE
                        song
                    }
                    searchResult.nextPager = result.nextPageToken
                    searchResult.songs = songs
                    searchResult.duration = System.currentTimeMillis() - startTime
                    success(searchResult)
                    mView?.hideLoading()
                }
                if (result == null) {
                    fail.invoke()
                }
            }
        }
    }

    override fun searchByType(key: String, offset: Int, type: Int) {
        ApiManager.request(NeteaseApiServiceImpl.searchMoreInfo(key, 10, offset, type),
                object : RequestCallBack<SearchInfo> {
                    override fun success(result: SearchInfo) {
                        when (type) {
                            1004 -> {

                            }
                            100 -> {

                            }
                            1000 -> {

                            }
                        }
                    }

                    override fun error(msg: String) {
                    }
                })
    }

    override fun searchLocal(key: String) {
        mView?.showLoading()
        doAsync {
            val result = DaoLitepal.searchLocalMusic(key)
            uiThread {
                mView?.hideLoading()
                mView?.showSearchResult(result)
            }
        }
    }

    override fun search(key: String, type: SearchEngine.Filter, limit: Int, page: Int) {
        LogUtil.e("SearchPresenter", "key :$key type:$type limit:$limit page:$page")
        mView?.showLoading()
        val observable = when (type) {
            SearchEngine.Filter.BAIDU -> BaiduApiServiceImpl.searchBaiduMusic(key, limit, page)
            else -> MusicApiServiceImpl.searchMusic(key, type, limit, page)
        }
        ApiManager.request(observable,
                object : RequestCallBack<MutableList<Music>> {
                    override fun success(result: MutableList<Music>) {
                        LogUtil.e("searchSuccess", result.toString())
                        mView?.hideLoading()
                        mView?.showSearchResult(result)
                    }

                    override fun error(msg: String) {
                        LogUtil.e("searchFail", msg)
                        mView?.hideLoading()
                        mView?.showSearchResult(mutableListOf())
                    }
                })
    }

    override fun getHotSearchInfo() {
        if (MusicApp.hotSearchList != null) {
            mView?.showHotSearchInfo(MusicApp.hotSearchList)
            return
        } else {
            ApiManager.request(NeteaseApiServiceImpl.getHotSearchInfo(), object : RequestCallBack<MutableList<HotSearchBean>> {
                override fun success(result: MutableList<HotSearchBean>?) {
                    result?.let {
                        MusicApp.hotSearchList = it
                        mView?.showHotSearchInfo(it)
                    }
                }

                override fun error(msg: String?) {
                }
            })
        }
    }

    override fun getSearchHistory() {
        doAsync {
            val data = DaoLitepal.getAllSearchInfo()
            uiThread {
                mView?.showSearchHistory(data)
            }
        }
    }

    override fun saveQueryInfo(query: String) {
        DaoLitepal.addSearchInfo(query)
    }
}
